src/Ocelot/Configuration/Repository/DiskFileConfigurationRepository.cs

```diff
@@ -14,21 +15,33 @@ namespace Ocelot.Configuration.Repository
     public class DiskFileConfigurationRepository : IFileConfigurationRepository
     {
         private readonly IOcelotConfigurationChangeTokenSource _changeTokenSource;
+        private readonly IOcelotCache<FileConfiguration> _cache;
         private readonly string _environmentFilePath;
         private readonly string _ocelotFilePath;
+        private readonly string _cacheKey;
         private static readonly object _lock = new();
         private const string ConfigurationFileName = "ocelot";
 
-        public DiskFileConfigurationRepository(IWebHostEnvironment hostingEnvironment, IOcelotConfigurationChangeTokenSource changeTokenSource)
+        public DiskFileConfigurationRepository(IWebHostEnvironment hostingEnvironment,
+            IOcelotConfigurationChangeTokenSource changeTokenSource, Cache.IOcelotCache<FileConfiguration> cache)
         {
             _changeTokenSource = changeTokenSource;
+            _cache = cache;
             _environmentFilePath = $"{AppContext.BaseDirectory}{ConfigurationFileName}{(string.IsNullOrEmpty(hostingEnvironment.EnvironmentName) ? string.Empty : ".")}{hostingEnvironment.EnvironmentName}.json";
+            _cacheKey = "InternalDiskFileConfigurationRepository";
 
             _ocelotFilePath = $"{AppContext.BaseDirectory}{ConfigurationFileName}.json";
         }
 
         public Task<Response<FileConfiguration>> Get()
         {
+            var configuration = _cache.Get(_cacheKey, _cacheKey);
+
+            if (configuration != null)
+            {
+                return Task.FromResult<Response<FileConfiguration>>(new OkResponse<FileConfiguration>(configuration));
+            }
+
             string jsonConfiguration;
 
             lock (_lock)
```

Comment:
Lock in a multithread context is not recommended.
Use SemaphoreSlim instead of

Reply:
Ray, it is not his design! 😉 
Truly speaking, I wonder, why does this method return a `Task<Response<FileConfiguration>>` object instead of `Response<FileConfiguration>` one? The method contains synchronous code only.

And the `File.ReadAllText` is synchronous, so we don't `await` this operation, so that's why it's fine to `lock` this operation. 
The [await doc](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/await) says: 
> Within an async method, you can't use the `await` operator in the body of a synchronous function, **inside the block of a [lock statement](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock)**, and in an [unsafe](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/unsafe) context.

And what is an argument to switch to `SemaphoreSlim`?
The `lock` (aka [Monitor](https://learn.microsoft.com/en-us/dotnet/api/system.threading.monitor?view=net-7.0)) has been used for ages, since [.NET 1.1](https://learn.microsoft.com/en-us/dotnet/api/system.threading.monitor?view=net-7.0#applies-to). 😄

Reply:
Agree with you, Raman, it is clearly not a async code, and in this context, lock is appropriate.
In a full async context, lock generates a compiler error (CS2996); there is a lot of articles who explain how to use SemaphoreSlim instead of lock in async context

Reply:
Not an issue, right?